File: src/web/web_app/src/lib/services/positioning/arrows/calculation/ArrowAdjustmentCalculator.ts
/**
 * Arrow Adjustment Calculator - Enhanced Orchestration Service
 *
 * Clean, focused coordinator service that delegates to specialized components.
 * Direct TypeScript port of the Python ArrowAdjustmentCalculator.
 *
 * ARCHITECTURE:
 * - DefaultPlacementService: Handles default placement lookups
 * - SpecialPlacementService: Handles special placement lookups
 * - DirectionalTupleProcessor: Handles tuple generation and selection
 * - This service: Coordinates the pipeline with proper error propagation
 *
 * USAGE:
 *     calculator = new ArrowAdjustmentCalculator(placementServices, tupleProcessor)
 *     adjustment = await calculator.calculateAdjustment(pictographData, motion, letter, location)
 */

import type { MotionData, PictographData } from '$lib/domain';
import { MotionType } from '$lib/domain';
import type { IArrowAdjustmentCalculator } from '../../core-services';
import type { Location, Point } from '../../types';
import { AttributeKeyGenerator } from '../key_generators/AttributeKeyGenerator';
import { PlacementKeyGenerator } from '../key_generators/PlacementKeyGenerator';
import { SpecialPlacementOriKeyGenerator } from '../key_generators/SpecialPlacementOriKeyGenerator';
import { TurnsTupleKeyGenerator } from '../key_generators/TurnsTupleKeyGenerator';
import { ArrowAdjustmentLookup as AdvancedLookup } from '../orchestration/ArrowAdjustmentLookup';
import { DefaultPlacementService } from '../placement/DefaultPlacementService';
import { SpecialPlacementService } from '../placement/SpecialPlacementService';
import {
	DirectionalTupleCalculator,
	DirectionalTupleProcessor,
	QuadrantIndexCalculator,
	type IDirectionalTupleProcessor,
} from '../processors/DirectionalTupleProcessor';

export type IArrowAdjustmentLookup = AdvancedLookup;

export class ArrowAdjustmentCalculator implements IArrowAdjustmentCalculator {
	/**
	 * Clean coordinator service for arrow positioning with proper error handling.
	 *
	 * Delegates to focused services:
	 * - ArrowAdjustmentLookup: Special/default placement lookups
	 * - DirectionalTupleProcessor: Tuple generation and selection
	 *
	 * Provides comprehensive arrow positioning adjustment calculation.
	 */

	private lookupService: IArrowAdjustmentLookup;
	private tupleProcessor: IDirectionalTupleProcessor;

	constructor(
		lookupService?: IArrowAdjustmentLookup,
		tupleProcessor?: IDirectionalTupleProcessor
	) {
		/**
		 * Initialize with focused services.
		 *
		 * Args:
		 *     lookupService: Service for adjustment lookups
		 *     tupleProcessor: Service for directional tuple processing
		 */
		// Use provided services or create with default dependencies
		this.lookupService = lookupService || this.createDefaultLookupService();
		this.tupleProcessor = tupleProcessor || this.createDefaultTupleProcessor();
	}

	async calculateAdjustment(
		pictographData: PictographData,
		motionData: MotionData,
		letter: string,
		location: Location,
		arrowColor?: string
	): Promise<Point> {
		/**
		 * Calculate arrow position adjustment with streamlined parameters.
		 *
		 * Args:
		 *     pictographData: Pictograph data containing context
		 *     motionData: Motion data containing type, rotation, and location info
		 *     letter: Letter for special placement lookup
		 *     location: Pre-calculated arrow location
		 *     arrowColor: Color of the arrow ('red' or 'blue')
		 *
		 * Returns:
		 *     Final position adjustment as Point (to be added to initial position)
		 */
		try {
			return await this.calculateAdjustmentResult(
				pictographData,
				motionData,
				letter,
				location,
				arrowColor
			);
		} catch (error) {
			// Log error and return default for backward compatibility
			console.error(`Adjustment calculation failed: ${error}`);
			return { x: 0, y: 0 };
		}
	}

	async calculateAdjustmentResult(
		pictographData: PictographData,
		motionData: MotionData,
		letter: string,
		location: Location,
		arrowColor?: string
	): Promise<Point> {
		/**
		 * Calculate arrow position adjustment with proper error handling.
		 *
		 * Args:
		 *     pictographData: Pictograph data containing context
		 *     motionData: Motion data containing type, rotation, and location info
		 *     letter: Letter for special placement lookup
		 *     location: Pre-calculated arrow location
		 *     arrowColor: Color of the arrow ('red' or 'blue')
		 *
		 * Returns:
		 *     Point adjustment
		 *
		 * Throws:
		 *     Error: If calculation fails due to invalid input or system error
		 */
		try {
			// STEP 1: Look up base adjustment (special → default) - EXACTLY like legacy
			const baseAdjustment = await this.lookupService.getBaseAdjustment(
				pictographData,
				motionData,
				letter,
				arrowColor
			);

			// STEP 2: Process directional tuples - EXACTLY like legacy
			const finalAdjustment = this.tupleProcessor.processDirectionalTuples(
				baseAdjustment,
				motionData,
				location
			);

			return finalAdjustment;
		} catch (error) {
			console.error(`Adjustment calculation failed for letter ${letter}: ${error}`);
			throw new Error(`Arrow adjustment calculation failed: ${error}`);
		}
	}

	calculateAdjustmentSync(
		pictographData: PictographData,
		motionData: MotionData,
		letter: string,
		location: Location,
		arrowColor?: string
	): Point {
		/**
		 * Synchronous version of calculateAdjustment for use in sync contexts.
		 * Note: This may have limited functionality compared to the async version.
		 */
		try {
			// For now, we'll use a simplified synchronous approach
			// In a full implementation, this would need synchronous versions of all services

			// Use the directional tuple processor directly with basic adjustments
			const baseAdjustment = this.getBasicAdjustmentSync(motionData);
			const finalAdjustment = this.tupleProcessor.processDirectionalTuples(
				baseAdjustment,
				motionData,
				location
			);

			console.debug(
				`Sync adjustment for ${motionData.motion_type} ${motionData.turns} turns at ${location}: (${finalAdjustment.x}, ${finalAdjustment.y})`
			);

			return finalAdjustment;
		} catch (error) {
			console.warn(`Sync adjustment calculation failed for letter ${letter}: ${error}`);
			return { x: 0, y: 0 };
		}
	}

	private getBasicAdjustmentSync(motionData: MotionData): Point {
		/**Get basic adjustment values for synchronous operation using proper placement logic.*/
		const motionType = motionData.motion_type;
		const turns = typeof motionData.turns === 'number' ? motionData.turns : 0;
		const turnsStr = turns === Math.floor(turns) ? turns.toString() : turns.toString();

		// Use actual placement data structure (simplified for sync operation)
		// These values come from the actual JSON files in static/data/arrow_placement
		const placementData: Record<string, Record<string, [number, number]>> = {
			[MotionType.PRO]: {
				'0': [-10, -40],
				'0.5': [30, 105],
				'1': [30, 25],
				'1.5': [-35, 145],
				'2': [-10, -35],
				'2.5': [20, 100],
				'3': [30, 25],
			},
			[MotionType.ANTI]: {
				'0': [0, -40],
				'0.5': [-15, 110],
				'1': [0, -40],
				'1.5': [20, 155],
				'2': [0, -40],
				'2.5': [0, 100],
				'3': [0, -50],
			},
			[MotionType.STATIC]: {
				'0': [0, 0],
				'0.5': [0, -140],
				'1': [50, 50],
				'1.5': [0, 0],
				'2': [0, 0],
				'2.5': [0, 0],
				'3': [0, 0],
			},
			[MotionType.DASH]: {
				'0': [0, 0],
			},
			[MotionType.FLOAT]: {
				fl: [30, -30],
			},
		};

		const motionAdjustments = placementData[motionType];
		if (motionAdjustments && motionAdjustments[turnsStr]) {
			const [x, y] = motionAdjustments[turnsStr];
			return { x, y };
		}

		return { x: 0, y: 0 };
	}

	private createDefaultLookupService(): IArrowAdjustmentLookup {
		/**Create lookup service with default dependencies.*/
		return new AdvancedLookup(
			new SpecialPlacementService(),
			new DefaultPlacementService(),
			new SpecialPlacementOriKeyGenerator(),
			new PlacementKeyGenerator(),
			new TurnsTupleKeyGenerator(),
			new AttributeKeyGenerator()
		);
	}

	private createDefaultTupleProcessor(): IDirectionalTupleProcessor {
		/**Create tuple processor with default dependencies.*/
		return new DirectionalTupleProcessor(
			new DirectionalTupleCalculator(),
			new QuadrantIndexCalculator()
		);
	}
}

File: src/desktop/modern/application/services/positioning/arrows/orchestration/arrow_adjustment_calculator.py
"""
Arrow Adjustment Calculator Service - Refactored with Result Types

Clean, focused coordinator service that delegates to specialized components.
Implements proper error handling with Result types instead of silent failures.

ARCHITECTURE:
- ArrowAdjustmentLookupService: Handles special/default placement lookups
- DirectionalTupleProcessor: Handles tuple generation and selection
- This service: Coordinates the pipeline with proper error propagation

USAGE:
    calculator = ArrowAdjustmentCalculatorService(lookup_service, tuple_processor)
    try:
        adjustment = calculator.calculate_adjustment(pictograph_data, motion_data, letter, location)
    except Exception as e:
        logger.error(f"Calculation failed: {e}")
"""

import logging

# Import the focused services
from desktop.modern.application.services.positioning.arrows.orchestration.arrow_adjustment_lookup import (
    ArrowAdjustmentLookup,
)
from desktop.modern.core.interfaces.positioning_services import (
    IArrowAdjustmentCalculator,
)
from desktop.modern.core.types.coordinates import get_default_point
from desktop.modern.core.types.geometry import Point
from desktop.modern.domain.models.enums import Location
from desktop.modern.domain.models.motion_data import MotionData
from desktop.modern.domain.models.pictograph_data import PictographData

from ..calculation.directional_tuple_calculator import DirectionalTupleCalculator
from ..calculation.quadrant_index_calculator import QuadrantIndexCalculator
from ..key_generators.attribute_key_generator import AttributeKeyGenerator
from ..key_generators.placement_key_generator import PlacementKeyGenerator
from ..key_generators.turns_tuple_key_generator import TurnsTupleKeyGenerator

# Legacy service imports for backward compatibility
from ..placement.default_placement_service import DefaultPlacementService
from ..placement.special_placement_ori_key_generator import (
    SpecialPlacementOriKeyGenerator,
)
from ..placement.special_placement_service import SpecialPlacementService
from .directional_tuple_processor import DirectionalTupleProcessor

logger = logging.getLogger(__name__)


class ArrowAdjustmentCalculator(IArrowAdjustmentCalculator):
    """
    Clean coordinator service for arrow positioning with proper error handling.

    Delegates to focused services:
    - ArrowAdjustmentLookupService: Special/default placement lookups
    - DirectionalTupleProcessor: Tuple generation and selection

    Provides both new Result-based API and legacy Point-based API for compatibility.
    """

    def __init__(
        self,
        lookup_service: ArrowAdjustmentLookup = None,
        tuple_processor: DirectionalTupleProcessor = None,
    ):
        """
        Initialize with focused services.

        Args:
            lookup_service: Service for adjustment lookups
            tuple_processor: Service for directional tuple processing
        """
        # Use provided services or create with default dependencies
        if lookup_service is None:
            lookup_service = self._create_default_lookup_service()
        if tuple_processor is None:
            tuple_processor = self._create_default_tuple_processor()

        self.lookup_service = lookup_service
        self.tuple_processor = tuple_processor

    def _create_default_lookup_service(self) -> ArrowAdjustmentLookup:
        """Create lookup service with default dependencies."""
        return ArrowAdjustmentLookup(
            special_placement_service=SpecialPlacementService(),
            default_placement_service=DefaultPlacementService(),
            orientation_key_service=SpecialPlacementOriKeyGenerator(),
            placement_key_service=PlacementKeyGenerator(),
            turns_tuple_service=TurnsTupleKeyGenerator(),
            attribute_key_service=AttributeKeyGenerator(),
        )

    def _create_default_tuple_processor(self) -> DirectionalTupleProcessor:
        """Create tuple processor with default dependencies."""
        return DirectionalTupleProcessor(
            directional_tuple_service=DirectionalTupleCalculator(),
            quadrant_index_service=QuadrantIndexCalculator(),
        )

    def calculate_adjustment(
        self,
        pictograph_data: PictographData,
        motion_data: MotionData,
        letter: str,
        location: Location,
        arrow_color: str = None,
    ) -> Point:
        """
        Calculate arrow position adjustment with streamlined parameters.

        Args:
            motion_data: Motion data containing type, rotation, and location info
            letter: Letter for special placement lookup
            location: Pre-calculated arrow location
            arrow_color: Color of the arrow ('red' or 'blue')

        Returns:
            Final position adjustment as Point (to be added to initial position)
        """
        try:
            return self.calculate_adjustment_result(
                pictograph_data, motion_data, letter, location, arrow_color
            )
        except Exception as e:
            # Log error and return default for backward compatibility
            logger.error(f"Adjustment calculation failed: {e}")
            return get_default_point()

    def calculate_adjustment_result(
        self,
        pictograph_data: PictographData,
        motion_data: MotionData,
        letter: str,
        location: Location,
        arrow_color: str = None,
    ) -> Point:
        """
        Calculate arrow position adjustment with proper error handling.

        Args:
            motion_data: Motion data containing type, rotation, and location info
            letter: Letter for special placement lookup
            location: Pre-calculated arrow location
            arrow_color: Color of the arrow ('red' or 'blue')

        Returns:
            Point adjustment

        Raises:
            ValueError: If calculation fails due to invalid input
            RuntimeError: If calculation fails due to system error
        """
        try:
            # STEP 1: Look up base adjustment (special → default) - EXACTLY like legacy
            base_adjustment = self.lookup_service.get_base_adjustment(
                pictograph_data, motion_data, letter, arrow_color
            )

            # STEP 2: Process directional tuples - EXACTLY like legacy
            final_adjustment = self.tuple_processor.process_directional_tuples(
                base_adjustment, motion_data, location
            )

            return final_adjustment

        except Exception as e:
            logger.error(f"Adjustment calculation failed for letter {letter}: {e}")
            raise RuntimeError(f"Arrow adjustment calculation failed: {e}") from e
